polyply/src/meta_molecule.py

```diff
@@ -105,39 +112,56 @@ def add_monomer(self, current, resname, connections):
-
-        self.old_resids = nx.get_node_attributes(self.molecule, "resid")
-        self.old_resnames = nx.get_node_attributes(self.molecule, "resname")
-
-        for node in self.nodes:
-            old_resid = self.nodes[node]["resid"]
-            if self.nodes[node]["resname"] == resname:
-                old_atoms = find_atoms(self.molecule, "resid", old_resid)
-                atom_names = {self.molecule.nodes[atom]["atomname"]:atom for atom in old_atoms}
-                for new_res in new_residues:
-                    new_name, atoms = new_res.split("-")
-                    names = atoms.split(",")
-                    for atom in names:
-                        try:
-                            node_key = atom_names[atom]
-                        except KeyError:
-                            msg = ("Residue {} {} has no atom {}.")
-                            raise IOError(msg.format(resname, old_resid, atom))
-                        self.molecule.nodes[node_key]["resname"] = new_name
-                        self.molecule.nodes[node_key]["resid"] = max_resid + 1
-                    max_resid += 1
+    def relable_and_redo_res_graph(self, mapping):
+        """
+        Relable the nodes of `self.molecule` using `mapping`
+        and regenerate the meta_molecule (i.e. residue graph).
 
+        Parameters:
+        -----------
+        mapping: dict
+            mapping of node-key to new residue name
+        """
+        # find the maxium resiude id
+        max_resid = max(nx.get_node_attributes(self.molecule, "resid").values())
+        # resname the residues and increase with pseudo-resid
+        for node, resname in mapping.items():
+            self.molecule.nodes[node]["resname"] = resname
+            old_resid = self.molecule.nodes[node]["resid"]
+            self.molecule.nodes[node]["resid"] = old_resid + max_resid + 1
+
+        # make a new residue graph and overwrite the old one
         new_meta_graph = make_residue_graph(self.molecule, attrs=('resid', 'resname'))
         self.clear()
         self.add_nodes_from(new_meta_graph.nodes(data=True))
         self.add_edges_from(new_meta_graph.edges)
-        return max_resid
+
+    def split_residue(self, split_strings):
+        """
+        Split all residues defind by the in `split_strings`, which is a list
+        of strings with format <resname>:<new_resname>-<atom1>,<atom2><etc>
+        into new residues and also update the underlying molecule with these
+        new residues.
+
+        Parameters:
+        -----------
+        split_strings:  list[str]
+             list of split strings
+
+        Returns:
+        dict
+            mapping of the nodes to the new resnames
+        """
+        mapping = {}
+        for split_string in split_strings:
+            # split resname and new resiude definitions
+            resname = split_string.split(":")[0]
+            new_residues = split_string.split(":")[1:]
```

Comment:
```suggestion
            resname, *new_residues = split_string.split(":")
```

Reply:
yes, still getting used to this syntax

Reply:
Oh, you may want to add a "max splits" to the split:
`resname, *new_residues = split_string.split(":", 1)`